### PR TITLE
mounts: Ignore existing mounts if they cannot be honored

### LINKED ISSUE #1068: mounts: Ignore existing mounts if they cannot be honored
In case we use an hypervisor that cannot support filesystem sharing, we copy files over to the VM rootfs through the gRPC protocol.
This is a nice workaround, but it only works with regular files, which means no device file, no socket file, no directory, etc... can be sent this way.
That's why we should simply accept the limitation in this use case, and ignore any non-regular file.

### PR DESCRIPTION
**Pull request by sboeuf**
In case we use an hypervisor that cannot support filesystem sharing,
we copy files over to the VM rootfs through the gRPC protocol. This
is a nice workaround, but it only works with regular files, which
means no device file, no socket file, no directory, etc... can be
sent this way.

This is a limitation that we accept here, by simply ignoring those
non-regular files.

Fixes #1068

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

**Comment by sboeuf** [comment]
/cc @mcastelino @egernst @bergwolf 
That's the fix to get k8s running with Kata+FC :smiley:

File: virtcontainers/container.go [hunk 1]
```diff
@@ -445,29 +445,45 @@ func (c *Container) shareFiles(m Mount, idx int, hostSharedDir, guestSharedDir s
...
+		if err != nil {
+			return "", false, err
+		}
+
+		// Ignore the mount if this is not a regular file (excludes
+		// directory, socket, device, ...) as it cannot be handled by
+		// a simple copy. But this should not be treated as an error,
+		// only as a limitation.
+		if !fileInfo.Mode().IsRegular() {
+			c.Logger().Debug("Ignore (%s), fs sharing not supported and not a regular file", m.Source)
```
Comment (jodh-intel) [inline virtcontainers/container.go]: Nit: this and other log calls could be converted into structured format:

```go
c.Logger().WithField("ignored-file", m.Source).Debug("Ignoring non-regular file as FS sharing not supported")
```

Comment (bergwolf) [reply virtcontainers/container.go]: And it is causing CI failure:
```
# github.com/kata-containers/runtime/virtcontainers
virtcontainers/container.go:459: Entry.Debug call has possible formatting directive %s
FAIL	github.com/kata-containers/runtime/virtcontainers [build failed]
Makefile:421: recipe for target 'go-test' failed
```

Comment (egernst) [reply virtcontainers/container.go]: I fixed it for him and repushed.

File: virtcontainers/kata_agent.go [hunk 2]
```diff
@@ -719,6 +719,30 @@ func (k *kataAgent) replaceOCIMountSource(spec *specs.Spec, guestMounts []Mount)
...
 }
 
+func (k *kataAgent) removeIgnoredOCIMount(spec *specs.Spec, ignoredMounts []Mount) error {
+	var mounts []specs.Mount
+
+	for _, m := range spec.Mounts {
+		found := false
+		for _, ignoredMount := range ignoredMounts {
+			if ignoredMount.Source == m.Source {
+				k.Logger().Debugf("Removing OCI mount (%s)", m.Source)
```
Comment (jodh-intel) [inline virtcontainers/kata_agent.go]: Nit: this one could be converted to a structured format too.

Comment (egernst) [reply virtcontainers/kata_agent.go]: ooooooh.  ok, fixed!

Comment (jodh-intel) [reply virtcontainers/kata_agent.go]: Ta! ;)

File: virtcontainers/container.go [hunk 3]
```diff
@@ -431,10 +431,10 @@ func (c *Container) createContainersDirs() error {
 	return nil
 }
 
-func (c *Container) shareFiles(m Mount, idx int, hostSharedDir, guestSharedDir string) (string, error) {
+func (c *Container) shareFiles(m Mount, idx int, hostSharedDir, guestSharedDir string) (string, bool, error) {
```
Comment (mcastelino) [inline virtcontainers/container.go]: @sboeuf it would be good to document what the return values actually mean or give them expressive names.

File: virtcontainers/container.go [hunk 4]
```diff
@@ -445,29 +445,45 @@ func (c *Container) shareFiles(m Mount, idx int, hostSharedDir, guestSharedDir s
...
+
+		// Ignore the mount if this is not a regular file (excludes
+		// directory, socket, device, ...) as it cannot be handled by
+		// a simple copy. But this should not be treated as an error,
+		// only as a limitation.
+		if !fileInfo.Mode().IsRegular() {
+			c.Logger().WithField("ignored-file", m.Source).Debug("Ignoring non-regular file as FS sharing not supported")
+			return "", true, nil
+		}
+
```
Comment (mcastelino) [inline virtcontainers/container.go]: @sboeuf we need to monitor these files on the host side to ensure that if they get updated we push the changes in. Maybe raise an issue

File: virtcontainers/container.go [hunk 5]
```diff
@@ -445,29 +445,45 @@ func (c *Container) shareFiles(m Mount, idx int, hostSharedDir, guestSharedDir s
...
+	return guestDest, false, nil
 }
 
 // mountSharedDirMounts handles bind-mounts by bindmounting to the host shared
 // directory which is mounted through 9pfs in the VM.
 // It also updates the container mount list with the HostPath info, and store
 // container mounts to the storage. This way, we will have the HostPath info
 // available when we will need to unmount those mounts.
-func (c *Container) mountSharedDirMounts(hostSharedDir, guestSharedDir string) ([]Mount, error) {
+func (c *Container) mountSharedDirMounts(hostSharedDir, guestSharedDir string) ([]Mount, []Mount, error) {
```
Comment (mcastelino) [inline virtcontainers/container.go]: The comments need to be updated to reflect the new logic and that the ignored mounts actually means.
Also we special case /dev/shm today.

File: virtcontainers/kata_agent.go [hunk 6]
```diff
@@ -1769,6 +1798,12 @@ func (k *kataAgent) copyFile(src, dst string) error {
 		Gid:      int32(st.Gid),
 	}
 
+	// Handle the special case where the file is empty
```
Comment (mcastelino) [inline virtcontainers/kata_agent.go]: @sboeuf why was this needed?

Comment (sboeuf) [reply virtcontainers/kata_agent.go]: Because the file was not copied at all when the size was 0, because of the logic below.

File: virtcontainers/container.go (hunk 4 above)
Comment (egernst) [reply virtcontainers/container.go]: (open issue, and mark it as known limitation)

**Comment by raravena80** [comment]
@sboeuf  Q. Will this work with Kubernetes `ConfigMaps`?

**Comment by mcastelino** [comment]
> @sboeuf Q. Will this work with Kubernetes `ConfigMaps`?
> 
> Meaning the `ConfigMap` will be copied to the firecracker VM?

@raravena80 it will not. I am opening an issue with details the limitations.
Today we copy files only in the case of pure files.

Configmap, Secrets. Tokens etc as shared as directories with keys set to file names and values the contents. We need to enhance our logic to handle these as a next step.

I am putting together a proposal for the same on special casing to handle a few critical features.

**Comment by mcastelino** [comment]
@raravena80 https://github.com/kata-containers/runtime/issues/1071
